_posts/2025-03-20-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,637 @@
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: strimzi-cluster-operator
+          ...
+          env:
+          ...
+          - name: STRIMZI_KAFKA_IMAGES
+            value: |
+              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
+              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
+              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
+          - name: STRIMZI_KAFKA_CONNECT_IMAGES
+            value: |
+              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
+              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
+              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
+          - name: STRIMZI_KAFKA_MIRROR_MAKER_IMAGES
+            value: |
+              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
+              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
+              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
+          - name: STRIMZI_KAFKA_MIRROR_MAKER_2_IMAGES
+            value: |
+              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
+              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
+              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
+        ...
+```
+
+In the sample above you can see that that Kafka version maps to an image which contains the Strimzi version. 
+The container images used by Strimzi include scripts, health endpoint servers and other services which form an API for the container.
+A given version of Strimzi expects certain endpoints, scripts etc to be present on the container in order to perform its operations. 
+This is why, even if the Kafka version has not changed, the image must change when you upgrade Strimzi.
+
+The result of this is that when you upgrade the Cluster Operator it will, as part of the reconciliation, change the image value on **all** of the Kafka broker pods in all of the Kafka clusters managed by that Strimzi instance on the Kubernetes cluster. 
+This means that all of your Kafka fleet (managed by that operator instance) will need to roll. 
+Strimzi will do this one broker at time and maintain topic availability (provided your topics have a replication factor greater than 1).
+However, rolling every Kafka cluster in your fleet could be quite disruptive.
+
+So what can we do to add more control?
+
+## Phased upgrades
+
+What we need is a way to control which Kafka clusters get upgraded and in what order. 
+This is often referred to as a phased upgrade or as an upgrade with _"blast radius"_ control because we can limit the application of the upgrade to a small number of resources (the radius) and contain any potential negative effects (the blast).
+
+There are several possible ways to implement phased upgrades of the Kafka clusters managed by a Strimzi operator:
+
+### Custom image per Kafka cluster
```

Comment:
This would IMHO not work for most versions and has possibly many different fallouts. You should at least remove this section if not even mention it explicitly as something to absolutely avoid!

Reply:
Oh I absolutely agree, I only mentioned it as I saw this proposal talked about seriously in the wild and wanted to explain why it was a bad idea.

But removing it seems reasonable, it is not a legitimate option and the post is already quite long.